numpy/_core/src/multiarray/stringdtype/casts.cpp

```diff
@@ -19,36 +23,243 @@
+        case NPY_CFLOAT:
+            return NPY_CFLOATLTR;
+        case NPY_CDOUBLE:
+            return NPY_CDOUBLELTR;
+        case NPY_CLONGDOUBLE:
+            return NPY_CLONGDOUBLELTR;
+        case NPY_OBJECT:
+            return NPY_OBJECTLTR;
+        case NPY_STRING:
+            return NPY_STRINGLTR;
+        case NPY_UNICODE:
+            return NPY_UNICODELTR;
+        case NPY_VOID:
+            return NPY_VOIDLTR;
+        case NPY_DATETIME:
+            return NPY_DATETIMELTR;
+        case NPY_TIMEDELTA:
+            return NPY_TIMEDELTALTR;
+        case NPY_CHAR:
+            return NPY_CHARLTR;
+        case NPY_VSTRING:
+            return NPY_VSTRINGLTR;
+        default:
+            PyErr_SetString(
+                PyExc_TypeError,
+                "No NPY_TYPECHAR associated with the given typenum."
+            );
+            return static_cast<NPY_TYPECHAR>('\0');
+    }
+}
+
+// Get a c string representation of a type number.
+const char *typenum_to_cstr(NPY_TYPES typenum) {
+    switch (typenum) {
+        case NPY_BOOL:
+            return "bool";
+        case NPY_BYTE:
+            return "byte";
+        case NPY_UBYTE:
+            return "ubyte";
+        case NPY_SHORT:
+            return "short";
+        case NPY_USHORT:
+            return "ushort";
+        case NPY_INT:
+            return "int";
+        case NPY_UINT:
+            return "uint";
+        case NPY_LONG:
+            return "long";
+        case NPY_ULONG:
+            return "ulong";
+        case NPY_LONGLONG:
+            return "long long";
+        case NPY_ULONGLONG:
+            return "ulong long";
+        case NPY_HALF:
+            return "half";
+        case NPY_FLOAT:
+            return "float";
+        case NPY_DOUBLE:
+            return "double";
+        case NPY_LONGDOUBLE:
+            return "long double";
+        case NPY_CFLOAT:
+            return "complex float";
+        case NPY_CDOUBLE:
+            return "complex double";
+        case NPY_CLONGDOUBLE:
+            return "complex long double";
+        case NPY_OBJECT:
+            return "object";
+        case NPY_STRING:
+            return "string";
+        case NPY_UNICODE:
+            return "unicode";
+        case NPY_VOID:
+            return "void";
+        case NPY_DATETIME:
+            return "datetime";
+        case NPY_TIMEDELTA:
+            return "timedelta";
+        case NPY_CHAR:
+            return "char";
+        case NPY_NOTYPE:
+            return "no type";
+        case NPY_USERDEF:
+            return "user defined";
+        case NPY_VSTRING:
+            return "vstring";
+        default:
+            return "unknown";
+    }
+}
+
+// Get the pointer to the PyArray_DTypeMeta for the type associated with the typenum.
+//
+// Returns NULL if the given typenum has no corresponding PyArray_DTypeMeta object;
+// error handling is left to the caller.
+PyArray_DTypeMeta *typenum_to_dtypemeta(NPY_TYPES typenum) {
```

Comment:
Same issue with `static` as above.

I'm also pretty sure you can do this without a switch statement:

```
NPY_DTYPE(PyArray_DescrFromType(typenum))
```

Reply:
Ahh, much better - thank you!